src/tir/schedule/primitive/compute_at.cc

```diff
@@ -0,0 +1,584 @@
+    }
+    return StmtMutator::VisitStmt_(loop);
+  }
+
+ public:
+  /*! \brief The root block of the block scope */
+  Block scope_root_;
+  /*! \brief The given block to be moved */
+  Block block_;
+  /*! \brief The given loop the block and its loop nest to be put under */
+  For loop_;
+  /*! \brief The new loop to replace the original loop */
+  For new_loop_{nullptr};
+  /*! \brief The new block realize to the moved block */
+  BlockRealize new_block_realize_{nullptr};
+  /*! \brief The plan to remove the given block by replacing this loop/block in the AST */
+  Stmt rm_src_stmt_{nullptr};
+  /*! \brief The plan to remove the given block by replacing to this loop/block in the AST */
+  Stmt rm_tgt_stmt_{nullptr};
+};
+
+/*!
+ * \brief Calculate a list of accessed buffer regions under a path of loops
+ * \tparam relax_storage_scope Whether to relax beyond the path according to the storage and
+ * execution scope
+ * \param binding The block binding, used to unbind the buffer regions
+ * \param buffer_regions The buffer regions to be calculated
+ * \param relax_path_low_inclusive The lowest point in the loop path, inclusive
+ * \param relax_path_high_exclusive The highest point in the loop path, exclusive
+ * \param relaxed Where the calculation result is stored
+ */
+template <bool relax_storage_scope>
+void RelaxBufferRegions(const Map<Var, PrimExpr>& binding,
+                        const Array<BufferRegion>& buffer_regions,
+                        const StmtSRef& relax_path_low_inclusive,
+                        const StmtSRef& relax_path_high_exclusive,
+                        std::unordered_map<const BufferNode*, std::vector<NDIntSet>>* relaxed) {
+  runtime::StorageScope global_scope{runtime::StorageRank::kGlobal, ""};
+  // We cache the variable domains
+  runtime::StorageRank previous_rank = runtime::StorageRank::kGlobal;
+  Optional<Map<Var, arith::IntSet>> var_dom = NullOpt;
+  // Enumerate every buffer region
+  for (const BufferRegion& buffer_region : buffer_regions) {
+    const Buffer& buffer = buffer_region->buffer;
+    const Array<Range>& region = buffer_region->region;
+    // Skip the buffer regions we are not interested in
+    auto it = relaxed->find(buffer.get());
+    if (it == relaxed->end()) {
+      continue;
+    }
+    std::vector<NDIntSet>& relaxed_regions = it->second;
+    // Check and update the cached `var_dom`
+    runtime::StorageScope scope =
+        relax_storage_scope ? runtime::StorageScope::Create(buffer.scope()) : global_scope;
+    runtime::StorageRank rank = scope.rank;
+    if (rank != previous_rank || !var_dom.defined()) {
+      previous_rank = rank;
+      var_dom = AsIntSet(LoopDomainOfSRefTreePath(
+          /*low_inclusive=*/relax_path_low_inclusive,
+          /*high_exclusive=*/relax_path_high_exclusive,
+          /*extra_relax_scope=*/scope));
+    }
+    // Relax the region
+    Array<arith::IntSet> relaxed = arith::EvalSet(Substitute(region, binding), var_dom.value());
```

Comment:
The name `relaxed` is duplicated with the parameter name. I think we'd better change a new name. Perhaps it can be `relaxed_region`.

Reply:
good idea